In [11]:
# 1. IMPORT REQUIRED LIBRARIES

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path

print("Libraries imported successfully!")

Libraries imported successfully!


In [12]:
# 2. DEFINE PROJECT PATHS

PROJECT_ROOT = Path.cwd()

# If the notebook is running from the notebooks/ folder,
# move one level up to the project root.
if PROJECT_ROOT.name == "notebooks":
    PROJECT_ROOT = PROJECT_ROOT.parent

DATA_DIR = PROJECT_ROOT / "data"
IBTRACS_DIR = DATA_DIR / "raw" / "ibtracs"

NI_FILE = IBTRACS_DIR / "ibtracs.NI.list.v04r01.csv"
GLOBAL_FILE = IBTRACS_DIR / "ibtracs.since1980.list.v04r01.csv"

print("Project root:", PROJECT_ROOT)
print("IBTrACS folder:", IBTRACS_DIR)
print()
print("NI file exists:", NI_FILE.exists())
print("Global file exists:", GLOBAL_FILE.exists())

Project root: d:\SIH\tropical-cyclone-ai
IBTrACS folder: d:\SIH\tropical-cyclone-ai\data\raw\ibtracs

NI file exists: True
Global file exists: True


In [14]:
# 3. CHECK FILE SIZES

def file_size_mb(path):
    return path.stat().st_size / (1024 ** 2)

print(f"NI dataset size: {file_size_mb(NI_FILE):.2f} MB")
print(f"Global dataset size: {file_size_mb(NI_FILE):.2f} MB")

NI dataset size: 26.59 MB
Global dataset size: 26.59 MB


In [15]:
# 4. INSPECT RAW CSV HEADER

with open(NI_FILE, 'r', encoding="utf-8", errors="replace") as f:
    for i in range(8):
        print(f.readline().rstrip())

SID,SEASON,NUMBER,BASIN,SUBBASIN,NAME,ISO_TIME,NATURE,LAT,LON,WMO_WIND,WMO_PRES,WMO_AGENCY,TRACK_TYPE,DIST2LAND,LANDFALL,IFLAG,USA_AGENCY,USA_ATCF_ID,USA_LAT,USA_LON,USA_RECORD,USA_STATUS,USA_WIND,USA_PRES,USA_SSHS,USA_R34_NE,USA_R34_SE,USA_R34_SW,USA_R34_NW,USA_R50_NE,USA_R50_SE,USA_R50_SW,USA_R50_NW,USA_R64_NE,USA_R64_SE,USA_R64_SW,USA_R64_NW,USA_POCI,USA_ROCI,USA_RMW,USA_EYE,TOKYO_LAT,TOKYO_LON,TOKYO_GRADE,TOKYO_WIND,TOKYO_PRES,TOKYO_R50_DIR,TOKYO_R50_LONG,TOKYO_R50_SHORT,TOKYO_R30_DIR,TOKYO_R30_LONG,TOKYO_R30_SHORT,TOKYO_LAND,CMA_LAT,CMA_LON,CMA_CAT,CMA_WIND,CMA_PRES,HKO_LAT,HKO_LON,HKO_CAT,HKO_WIND,HKO_PRES,KMA_LAT,KMA_LON,KMA_CAT,KMA_WIND,KMA_PRES,KMA_R50_DIR,KMA_R50_LONG,KMA_R50_SHORT,KMA_R30_DIR,KMA_R30_LONG,KMA_R30_SHORT,NEWDELHI_LAT,NEWDELHI_LON,NEWDELHI_GRADE,NEWDELHI_WIND,NEWDELHI_PRES,NEWDELHI_CI,NEWDELHI_DP,NEWDELHI_POCI,REUNION_LAT,REUNION_LON,REUNION_TYPE,REUNION_WIND,REUNION_PRES,REUNION_TNUM,REUNION_CI,REUNION_RMW,REUNION_R34_NE,REUNION_R34_SE,REUNION_R34_SW,REUNION_R

In [17]:
# 5. LOAD IBTRACS DATASETS

# Skip row 2 because it contains units, not actual data.
# Row numbering here starts from 0:
# row 0 = column names
# row 1 = units
# row 2 onward = actual cyclone records


ni_df = pd.read_csv(
    NI_FILE, skiprows=[1], low_memory=False
)

global_df = pd.read_csv(
    GLOBAL_FILE, skiprows=[1], low_memory=False
)

print("Datasets loaded successfully!")
print()
print("NI dataset shape: ", ni_df.shape)
print("Gloabal dataset shape: ", global_df.shape)

Datasets loaded successfully!

NI dataset shape:  (62848, 174)
Gloabal dataset shape:  (309202, 174)


In [18]:
# 6. PREVIEW THE DATA

print("NI Dataset:")
display(ni_df.head())

print("\nGlobal Dataset:")
display(global_df.head())

NI Dataset:


,SID,SEASON,NUMBER,BASIN,SUBBASIN,NAME,ISO_TIME,NATURE,LAT,LON,...,BOM_GUST_PER,REUNION_GUST,REUNION_GUST_PER,USA_SEAHGT,USA_SEARAD_NE,USA_SEARAD_SE,USA_SEARAD_SW,USA_SEARAD_NW,STORM_SPEED,STORM_DIR
0,1842298N11080,1842,1,NI,BB,UNNAMED,1842-10-25 03:00:00,NR,10.9,80.3,...,,,,,,,,,9,265
1,1842298N11080,1842,1,NI,BB,UNNAMED,1842-10-25 06:00:00,NR,10.9,79.8,...,,,,,,,,,9,265
2,1842298N11080,1842,1,NI,BB,UNNAMED,1842-10-25 09:00:00,NR,10.8,79.4,...,,,,,,,,,9,265
3,1842298N11080,1842,1,NI,BB,UNNAMED,1842-10-25 12:00:00,NR,10.8,78.9,...,,,,,,,,,9,265
4,1842298N11080,1842,1,NI,BB,UNNAMED,1842-10-25 15:00:00,NR,10.8,78.4,...,,,,,,,,,9,270



Global Dataset:


,SID,SEASON,NUMBER,BASIN,SUBBASIN,NAME,ISO_TIME,NATURE,LAT,LON,...,BOM_GUST_PER,REUNION_GUST,REUNION_GUST_PER,USA_SEAHGT,USA_SEARAD_NE,USA_SEARAD_SE,USA_SEARAD_SW,USA_SEARAD_NW,STORM_SPEED,STORM_DIR
0,1980001S13173,1980,1,SP,MM,PENI,1980-01-01 00:00:00,TS,-12.5,172.5,...,,,,,,,,,6,350
1,1980001S13173,1980,1,SP,MM,PENI,1980-01-01 03:00:00,TS,-12.2,172.4,...,,,,,,,,,6,350
2,1980001S13173,1980,1,SP,MM,PENI,1980-01-01 06:00:00,TS,-11.9,172.4,...,,,,,,,,,5,360
3,1980001S13173,1980,1,SP,MM,PENI,1980-01-01 09:00:00,TS,-11.7,172.4,...,,,,,,,,,4,10
4,1980001S13173,1980,1,SP,MM,PENI,1980-01-01 12:00:00,TS,-11.5,172.5,...,,,,,,,,,4,20


In [19]:
# 7. DATASET STRUCTURE

print("========== NI DATASET ==========")
print("Rows:", ni_df.shape[0])
print("Columns:", ni_df.shape[1])

print("\n========== GLOBAL DATASET ==========")
print("Rows:", global_df.shape[0])
print("Columns:", global_df.shape[1])

========== NI DATASET ==========
Rows: 62848
Columns: 174

========== GLOBAL DATASET ==========
Rows: 309202
Columns: 174


In [20]:
# 8. COLUMN NAMES

print("Total columns:", len(ni_df.columns))
print()

for i, column in enumerate(ni_df.columns, start=1):
    print(f"{i:3}. {column}")

Total columns: 174

  1. SID
  2. SEASON
  3. NUMBER
  4. BASIN
  5. SUBBASIN
  6. NAME
  7. ISO_TIME
  8. NATURE
  9. LAT
 10. LON
 11. WMO_WIND
 12. WMO_PRES
 13. WMO_AGENCY
 14. TRACK_TYPE
 15. DIST2LAND
 16. LANDFALL
 17. IFLAG
 18. USA_AGENCY
 19. USA_ATCF_ID
 20. USA_LAT
 21. USA_LON
 22. USA_RECORD
 23. USA_STATUS
 24. USA_WIND
 25. USA_PRES
 26. USA_SSHS
 27. USA_R34_NE
 28. USA_R34_SE
 29. USA_R34_SW
 30. USA_R34_NW
 31. USA_R50_NE
 32. USA_R50_SE
 33. USA_R50_SW
 34. USA_R50_NW
 35. USA_R64_NE
 36. USA_R64_SE
 37. USA_R64_SW
 38. USA_R64_NW
 39. USA_POCI
 40. USA_ROCI
 41. USA_RMW
 42. USA_EYE
 43. TOKYO_LAT
 44. TOKYO_LON
 45. TOKYO_GRADE
 46. TOKYO_WIND
 47. TOKYO_PRES
 48. TOKYO_R50_DIR
 49. TOKYO_R50_LONG
 50. TOKYO_R50_SHORT
 51. TOKYO_R30_DIR
 52. TOKYO_R30_LONG
 53. TOKYO_R30_SHORT
 54. TOKYO_LAND
 55. CMA_LAT
 56. CMA_LON
 57. CMA_CAT
 58. CMA_WIND
 59. CMA_PRES
 60. HKO_LAT
 61. HKO_LON
 62. HKO_CAT
 63. HKO_WIND
 64. HKO_PRES
 65. KMA_LAT
 66. KMA_LON
 67. KMA_CAT
 

In [21]:
# 9. DATA TYPES

ni_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 62848 entries, 0 to 62847
Columns: 174 entries, SID to STORM_DIR
dtypes: float64(2), int64(4), str(168)
memory usage: 83.4 MB


In [22]:
# 10. BASIC NUMERICAL STATISTICS

ni_df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
SID,62848,1858,1966233N13340,191,NaN,NaN,NaN,NaN,NaN,NaN,NaN
SEASON,62848.0,NaN,NaN,NaN,1953.215011,41.083059,1842.0,1922.0,1958.0,1986.0,2025.0
NUMBER,62848.0,NaN,NaN,NaN,57.247852,31.37591,1.0,31.0,51.0,84.0,140.0
BASIN,62366,2,NI,57841,NaN,NaN,NaN,NaN,NaN,NaN,NaN
SUBBASIN,62399,5,BB,42415,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...
USA_SEARAD_SE,62848,1,,62848,NaN,NaN,NaN,NaN,NaN,NaN,NaN
USA_SEARAD_SW,62848,1,,62848,NaN,NaN,NaN,NaN,NaN,NaN,NaN
USA_SEARAD_NW,62848,1,,62848,NaN,NaN,NaN,NaN,NaN,NaN,NaN
STORM_SPEED,62848,50,5,6589,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [24]:
# 11. MISSING VALUE ANALYSIS

missing = ni_df.isna().sum()

missing_summary = pd.DataFrame({
    "missing_count": missing,
    "missing_percentage": (missing / len(ni_df)) * 100
})

missing_summary = missing_summary.sort_values(
    by = "missing_percentage",
    ascending = False
)

display(missing_summary.head(30))

,missing_count,missing_percentage
BASIN,482,0.766930
SUBBASIN,449,0.714422
SEASON,0,0.000000
SID,0,0.000000
NUMBER,0,0.000000
NAME,0,0.000000
ISO_TIME,0,0.000000
NATURE,0,0.000000
LAT,0,0.000000
LON,0,0.000000


In [25]:
# 12. CHECK MISSING-VALUE REPRESENTATIONS


# Look at the most common values in a few important columns
# to see how missing information is represented.

columns_to_check = [
    "WMO_WIND",
    "WMO_PRES",
    "USA_WIND",
    "USA_PRES",
    "USA_SSHS",
    "USA_R34_NE",
    "USA_RMW"
]

for col in columns_to_check:
    print(f"\n--- {col} ---")
    print(ni_df[col].value_counts(dropna=False).head(10))


--- WMO_WIND ---
WMO_WIND
      54918
25     2171
30     1344
35      805
45      605
40      448
55      426
20      318
65      255
50      207
Name: count, dtype: int64

--- WMO_PRES ---
WMO_PRES
        54620
1000      821
1002      719
998       713
996       663
1004      618
994       558
990       556
992       435
988       275
Name: count, dtype: int64

--- USA_WIND ---
USA_WIND
      46486
25     2265
30     1693
35     1520
20     1366
45      882
40      817
15      798
50      605
55      565
Name: count, dtype: int64

--- USA_PRES ---
USA_PRES
        55738
1000      696
1004      650
1002      499
997       354
996       342
1007      338
1006      253
989       238
993       231
Name: count, dtype: int64

--- USA_SSHS ---
USA_SSHS
-5    46471
-1     6376
 0     5985
 1     1313
-3     1200
 2      536
 3      377
 4      287
-4      220
 5       83
Name: count, dtype: int64

--- USA_R34_NE ---
USA_R34_NE
       59893
60       208
70       177
55       172
50       167

In [26]:
# 13. COUNT WHITESPACE-ONLY VALUES

whitespace_summary = {}

for col in ni_df.columns:
    # Convert to string temporarily and check for blank/space-only values
    whitespace_count = ni_df[col].astype(str).str.strip().eq("").sum()
    whitespace_summary[col] = whitespace_count

whitespace_summary = pd.Series(whitespace_summary).sort_values(
    ascending=False
)

display(
    whitespace_summary.head(30).to_frame("whitespace_only_count")
)

,whitespace_only_count
NADI_LAT,62848
NADI_LON,62848
NADI_CAT,62848
NADI_WIND,62848
NADI_PRES,62848
WELLINGTON_LAT,62848
WELLINGTON_LON,62848
WELLINGTON_WIND,62848
WELLINGTON_PRES,62848
DS824_PRES,62848


In [29]:
# 14. CHECK SPECIAL / PLACEHOLDER VALUES

special_values = ["-999", "-999.0", "999", "NaN", "nan"]

for col in columns_to_check:
    counts = ni_df[col].astype(str).str.strip().isin(special_values).sum()

    if counts > 0:
        print(f"{col}: {counts} special/placeholder values")


WMO_PRES: 119 special/placeholder values
USA_PRES: 162 special/placeholder values


In [31]:
# 15. DUPLICATE RECORD ANALYSIS

duplicate_count = ni_df.duplicated().sum()

print("Exact duplicate rows: ", duplicate_count)
print("Percentage of duplicate rows: ", 
        round((duplicate_count / len(ni_df)) * 100, 4), "%")

Exact duplicate rows:  0
Percentage of duplicate rows:  0.0 %


In [36]:
# 16. CYCLONE / STORM ID ANALYSIS

unique_cyclones = ni_df["SID"].nunique()

print("Total reocrds:", len(ni_df))
print("Unique cyclone IDs:", unique_cyclones)
print("Average records per cyclone:", 
      round(len(ni_df) / unique_cyclones, 2))

Total reocrds: 62848
Unique cyclone IDs: 1858
Average records per cyclone: 33.83


In [38]:
# 17. OBSERVATIONS PER CYCLONE

observations_per_cyclone = (
    ni_df.groupby("SID")
    .size()
    .sort_values(ascending=False)
)

print("Cyclones with most observations:")
display(observations_per_cyclone.head(10))

print("\nCyclones with fewest observations:")
display(observations_per_cyclone.tail(10))

Cyclones with most observations:


SID
1966233N13340    191
1986228N19120    177
2013305N07141    159
1996288N09092    155
1932244N19296    145
1987219N08155    141
1979205N10148    137
1997298N06140    125
2006329N06150    123
1993336N05164    119
dtype: int64


Cyclones with fewest observations:


SID
1936177N21088    1
1936273N16095    1
1954211N19087    1
1954186N20088    1
1956161N22088    1
1960315N10081    1
1882212N20088    1
1907135N18092    1
1907238N21087    1
1932145N11075    1
dtype: int64

In [ ]:
# 18. DATE / TIME ANALYSIS

dates = pd.to_datetime(
    ni_df["ISO_TIME"],
    errors="coerce"
)
# Why errors="coerce"?
# If a date is invalid, pandas converts it to NaT instead of crashing.
# We're not modifying ni_df yet.

print("Eraliest observation:", dates.min())
print("Latest observation:", dates.max())
print("Invalid dates:", dates.isna().sum())



Eraliest observation: 1842-10-25 03:00:00
Latest observation: 2025-12-02 18:00:00
Invalid dates: 0


In [40]:
# 19. TIME INTERVAL ANALYSIS

time_differences = (
    ni_df.assign(parsed_time=dates)
    .sort_values(["SID", "parsed_time"])
    .groupby("SID")["parsed_time"]
    .diff()
)

print("Most common time intervals:")
display(time_differences.value_counts().head(10))

Most common time intervals:


parsed_time
0 days 03:00:00    60736
0 days 01:00:00      125
0 days 02:00:00      125
0 days 00:30:00        2
0 days 02:30:00        2
Name: count, dtype: int64

In [45]:
# 20. GEOGRAPHIC DATA CHECK (Check latitude and longitude)

lat = pd.to_numeric(ni_df["LAT"], errors="coerce")
lon = pd.to_numeric(ni_df["LON"], errors="coerce")

print("LAT")
print("  Minimum:", lat.min())
print("  Maximum:", lat.max())
print("  Invalid/missing:", lat.isna().sum())

print("\nLON")
print("  Minimum:", lon.min())
print("  Maximum:", lon.max())
print("  Invalid/missing:", lon.isna().sum())

LAT
  Minimum: 0.7
  Maximum: 83.0
  Invalid/missing: 0

LON
  Minimum: -87.7
  Maximum: 163.7
  Invalid/missing: 0


In [46]:
# 21. INVALID COORDINATE CHECK

invalid_lat = ((lat < -90) | (lat > 90)).sum()
invalid_lon = ((lon < -180) | (lon > 180)).sum()

print("Invalid latitude values:", invalid_lat)
print("Invalid longitude values:", invalid_lon)

Invalid latitude values: 0
Invalid longitude values: 0


In [47]:
# 22. IMPORTANT METEOROLOGICAL FIELDS

important_columns = [
    "WMO_WIND",
    "WMO_PRES",

    "USA_WIND",
    "USA_PRES",
    "USA_SSHS",
    "USA_RMW",

    "NEWDELHI_WIND",
    "NEWDELHI_PRES",
    "NEWDELHI_GRADE",

    "BOM_WIND",
    "BOM_PRES",

    "TOKYO_WIND",
    "TOKYO_PRES"
]

summary = []

for col in important_columns:
    if col in ni_df.columns:
        values = ni_df[col].astype(str).str.strip()

        # Count values that are not empty
        non_empty = values.ne("").sum()

        summary.append({
            "column": col,
            "non_empty": non_empty,
            "empty": len(ni_df) - non_empty,
            "data_percentage": round(
                non_empty / len(ni_df) * 100, 2
            )
        })

agency_summary = pd.DataFrame(summary)

display(
    agency_summary.sort_values(
        "data_percentage",
        ascending=False
    )
)

,column,non_empty,empty,data_percentage
4,USA_SSHS,62848,0,100.00
2,USA_WIND,16362,46486,26.03
8,NEWDELHI_GRADE,9558,53290,15.21
6,NEWDELHI_WIND,9467,53381,15.06
7,NEWDELHI_PRES,8605,54243,13.69
1,WMO_PRES,8228,54620,13.09
0,WMO_WIND,7930,54918,12.62
3,USA_PRES,7110,55738,11.31
5,USA_RMW,5617,57231,8.94
12,TOKYO_PRES,3225,59623,5.13


In [51]:
# 23. YEAR DISTRIBUTION

year_counts = (
    ni_df["SEASON"]
    .value_counts()
    .sort_index()
)

display(year_counts)


# Check how many records fall within the
# TCIR satellite dataset period: 2003–2016.

tcir_period = ni_df[
    ni_df["SEASON"].between(2003, 2016)
]

print(
    "NI records from 2003–2016:",
    len(tcir_period)
)

print(
    "Percentage of NI dataset:",
    round(len(tcir_period) / len(ni_df) * 100, 2),
    "%"
)

SEASON
1842     65
1845     25
1854     25
1877    234
1878    274
       ... 
2021    340
2022    413
2023    418
2024    402
2025    505
Name: count, Length: 152, dtype: int64

NI records from 2003–2016: 5103
Percentage of NI dataset: 8.12 %


In [50]:
# 24. UNIQUE CYCLONES IN TCIR PERIOD

tcir_cyclones = tcir_period["SID"].nunique()

print(
    "Unique NI cyclones from 2003–2016:",
    tcir_cyclones
)

Unique NI cyclones from 2003–2016: 132


In [52]:
# 25. COMPARE NI AND GLOBAL CYCLONE IDs

ni_sids = set(ni_df["SID"].unique())
global_sids = set(global_df["SID"].unique())

common_sids = ni_sids.intersection(global_sids)

print("Unique NI cyclone IDs:", len(ni_sids))
print("Unique Global cyclone IDs:", len(global_sids))
print("NI cyclone IDs also present in Global:", len(common_sids))

print(
    "Percentage of NI cyclone IDs found in Global:",
    round(len(common_sids) / len(ni_sids) * 100, 2),
    "%"
)

Unique NI cyclone IDs: 1858
Unique Global cyclone IDs: 4983
NI cyclone IDs also present in Global: 471
Percentage of NI cyclone IDs found in Global: 25.35 %


In [53]:
# 26. BASIN DISTRIBUTION

basin_counts = (
    ni_df["BASIN"]
    .astype(str)
    .str.strip()
    .value_counts()
)

display(basin_counts)

BASIN
NI    57841
WP     4525
Name: count, dtype: int64

In [54]:
# 27. SUB-BASIN DISTRIBUTION

subbasin_counts = (
    ni_df["SUBBASIN"]
    .astype(str)
    .str.strip()
    .value_counts()
)

display(subbasin_counts)

SUBBASIN
BB    42415
AS    15426
MM     4525
GM       25
CS        8
Name: count, dtype: int64

In [55]:
# 28. NATURE DISTRIBUTION

nature_counts = (
    ni_df["NATURE"]
    .astype(str)
    .str.strip()
    .value_counts()
)

display(nature_counts)

NATURE
TS    57157
NR     3377
DS     1271
MX      824
ET      219
Name: count, dtype: int64

In [56]:
# 29. INTENSITY / PRESSURE FIELD COMPLETENESS

intensity_columns = [
    "WMO_WIND",
    "WMO_PRES",
    "USA_WIND",
    "USA_PRES",
    "NEWDELHI_WIND",
    "NEWDELHI_PRES",
    "BOM_WIND",
    "BOM_PRES",
    "TOKYO_WIND",
    "TOKYO_PRES"
]

summary = []

for col in intensity_columns:
    values = ni_df[col].astype(str).str.strip()
    non_empty = values.ne("").sum()

    summary.append({
        "Field": col,
        "Available": non_empty,
        "Missing": len(ni_df) - non_empty,
        "Available_%": round(non_empty / len(ni_df) * 100, 2)
    })

intensity_summary = pd.DataFrame(summary)

display(
    intensity_summary.sort_values(
        "Available_%",
        ascending=False
    )
)

,Field,Available,Missing,Available_%
2,USA_WIND,16362,46486,26.03
4,NEWDELHI_WIND,9467,53381,15.06
5,NEWDELHI_PRES,8605,54243,13.69
1,WMO_PRES,8228,54620,13.09
0,WMO_WIND,7930,54918,12.62
3,USA_PRES,7110,55738,11.31
9,TOKYO_PRES,3225,59623,5.13
8,TOKYO_WIND,1260,61588,2.00
7,BOM_PRES,0,62848,0.00
6,BOM_WIND,0,62848,0.00


In [59]:
# 30. WIND SOURCE OVERLAP

wind_columns = [
    "USA_WIND",
    "NEWDELHI_WIND",
    "WMO_WIND",
    "TOKYO_WIND"
]

# Convert blank/whitespace values to NaN
wind_data = ni_df[wind_columns].copy()

for col in wind_columns:
    wind_data[col] = (
        wind_data[col]
        .astype(str)
        .str.strip()
        .replace("", np.nan)
    )

# Count how many wind sources are available
wind_data["available_sources"] = wind_data.notna().sum(axis=1)

print("Number of wind sources available per observation:")
display(
    wind_data["available_sources"]
    .value_counts()
    .sort_index()
)

print("\nObservations with at least one wind source:",
      (wind_data["available_sources"] >= 1).sum())

print("Observations with 2 or more wind sources:",
      (wind_data["available_sources"] >= 2).sum())

print("Observations with 3 or more wind sources:",
      (wind_data["available_sources"] >= 3).sum())

print("Observations with all 4 wind sources:",
      (wind_data["available_sources"] == 4).sum())

Number of wind sources available per observation:


available_sources
0    43240
1     9768
2     4275
3     5559
4        6
Name: count, dtype: int64


Observations with at least one wind source: 19608
Observations with 2 or more wind sources: 9840
Observations with 3 or more wind sources: 5565
Observations with all 4 wind sources: 6


In [60]:
# 31. ESSENTIAL TRACK FIELD QUALITY

essential_columns = [
    "SID",
    "ISO_TIME",
    "LAT",
    "LON"
]

quality_summary = []

for col in essential_columns:
    values = ni_df[col].astype(str).str.strip()

    quality_summary.append({
        "Field": col,
        "Missing": values.eq("").sum(),
        "Available": values.ne("").sum(),
        "Available_%": round(
            values.ne("").sum() / len(ni_df) * 100, 2
        )
    })

quality_df = pd.DataFrame(quality_summary)

display(quality_df)

,Field,Missing,Available,Available_%
0,SID,0,62848,100.0
1,ISO_TIME,0,62848,100.0
2,LAT,0,62848,100.0
3,LON,0,62848,100.0


In [61]:
# 32. TRACK TIME INTERVALS

parsed_time = pd.to_datetime(
    ni_df["ISO_TIME"],
    errors="coerce"
)

time_differences = (
    ni_df.assign(parsed_time=parsed_time)
    .sort_values(["SID", "parsed_time"])
    .groupby("SID")["parsed_time"]
    .diff()
)

print("Most common time intervals between observations:")

display(
    time_differences
    .value_counts()
    .head(10)
)

Most common time intervals between observations:


parsed_time
0 days 03:00:00    60736
0 days 01:00:00      125
0 days 02:00:00      125
0 days 00:30:00        2
0 days 02:30:00        2
Name: count, dtype: int64

# 01 Data Exploration — Summary

This notebook contains the initial exploration of the IBTrACS North Indian Ocean (NI) cyclone track dataset before data cleaning.

## Dataset
- 62,848 observations
- 174 columns
- 1,858 unique cyclones
- Data range: 1842–2025

## Data Quality
- No exact duplicate rows were found.
- `SID`, `ISO_TIME`, `LAT`, and `LON` are 100% available.
- Latitude and longitude values are within valid geographic ranges.
- No invalid dates were found.
- Most observations have a 3-hour time interval.

## Missing Data
- Many agency-specific fields contain blank values.
- Wind and pressure fields have significant missing data.
- `USA_WIND` has the highest availability among the checked wind fields at 26.03%.
- 19,608 observations (31.2%) have at least one available wind source.
- Missing intensity values will not be blindly filled or used as the only reason to remove track observations.

## TCIR Period
- TCIR covers 2003–2016.
- 5,103 NI observations belong to this period.
- These observations represent 132 unique cyclones.
- This period will be important when matching IBTrACS tracks with the TCIR satellite data.

## Conclusion
The core cyclone tracking fields (`SID`, `ISO_TIME`, `LAT`, and `LON`) are complete and reliable for further processing. However, intensity and agency-specific fields contain substantial missing data.

The raw IBTrACS files will remain unchanged. The next stage will clean and standardize the required track fields and save the cleaned data separately in `data/interim/cleaned_tracks/`.